03-rnaseq/clustering/clustering_heatmap_example.Rmd

```diff
@@ -0,0 +1,334 @@
+We are going to use the `rlog` function from the `DESeq2` package to normalize
+the data. 
+TODO: Add a note/link to section in documentation on why we use `rlog` vs `vst`
+and `normTransform`.
+
+```{r}
+# Normalize the data in the `DESeqDataSet` object using the
+# `rlog` function from the `DESEq2` R package
+dds_norm <- rlog(dds)
+```
+
+### Choose genes of interest
+Although you may want to create a heatmap including all of the genes in the set,
+alternatively, the heatmap could be created using only genes of interest. 
+For this example, we will sort genes by variance, but there are many alternative
+criterion by which you may want to sort your genes <i>eg</i> fold change,
+t-statistic, membership to a particular gene ontology, so on. 
+
+```{r}
+# Calculate the variance for each gene
+variances <- apply(assay(dds_norm), 1, var)
+
+# Determine the upper quartile variance cutoff value
+upper_var <- quantile(variances, 0.75)
+
+# Subset the data choosing only genes whose variances are in the upper quartile
+df_by_var <- data.frame(assay(dds_norm)) %>%
+  dplyr::filter(variances > upper_var)
+```
+
+### Create a heatmap
+To further customize the heatmap, see a vignette for a guide at this [link](https://slowkow.com/notes/pheatmap-tutorial/).
+
+```{r}
+# Create and store the technical replicates heatmap object
+pheatmap <-
+  pheatmap(
+    df_by_var,
+    cluster_rows = TRUE, # We want to cluster the heatmap by rows (genes in this case)
+    cluster_cols = TRUE, # We also want to cluster the heatmap by columns (samples in this case)
+    main = "Clustered Heatmap",
+    colorRampPalette(c(
+      "deepskyblue",
+      "black",
+      "yellow"
+    ))(25
+    ),
+    scale = "row" # Scale values in the direction of genes (rows)
+  )
+```
+
+We've created a heatmap but although our genes and samples are clustered, there
+is not much information that we can gather here because we did not provide the
+`pheatmap` function with annotation labels for our samples.
+
+First let's save our clustered heatmap.
+
+#### Save heatmap as a png
+You can easily switch this to save to a jpeg or tiff by changing the function 
+and file name within the function to the respective file suffix.
+
+```{r}
+# Open a png file
+png(file.path(
+  plots_dir,
+  "HeatmapSRP070849_clustered.png" # Replace file name with a relevant output plot name
+))
+
+# Print your heatmap
+pheatmap
+
+# Close the png file:
+dev.off()
+```
+
+Now, let's add some annotation bars to our heatmap.
+
+### Prepare metadata for annotation
+We are going to manipulate the metadata and add variables with the information
+that we would like to use to annotate the heatmap.
+
+```{r}
+# Let's prepare the annotation data.frame for the uncollapsed `DESeqData` set
+# object which will be used to create the technical replicates heatmap
+annotation_df <- metadata %>%
+  # Create a variable to store the cancer type information
+  dplyr::mutate(
+    cancer_type = dplyr::case_when(
```

Comment:
Since the mice don't have cancer per se, it might be a little clearer to name this `mutation` or something like that. (Again, haven't read the paper so you may want to double check this).

Reply:
Second. thought is if we are including both the `cancer_type` and `refinebio_treatment` as labels in the heatmap, the we should include them both in the model in the `DESeq2` object. (Even though we aren't using it for DE here but its just good to set a good set up ).

Reply:
> Second. thought is if we are including both the cancer_type and refinebio_treatment as labels in the heatmap, the we should include them both in the model in the DESeq2 object. (Even though we aren't using it for DE here but its just good to set a good set up ).

@cansavvy I get the following error when trying to implement this change: 
```
Error in checkFullRank(modelMatrix) : 
  the model matrix is not full rank, so the model cannot be fit as specified.
  One or more variables or interaction terms in the design formula are linear
  combinations of the others and must be removed.
```

Furthermore, before opening this revamped PR, @jaclyn-taroni suggested starting with the heatmap then preparing the annotation data.frame, noting that the metadata does not necessarily need to be cleaned up metadata for the `DESeqDataSet` object.

Reply:
Hmm.. Okay. Can you do a `table()` of those two variables? Perhaps there's not enough samples to use both variables, in which case, that's fine, we can just explain why we are choosing one over the other. 

I am a bit confused by this comment. Can you give me more context and/or a link?: 
> noting that the metadata does not necessarily need to be cleaned up metadata for the DESeqDataSet object.

Reply:
Also can you paste here the code or permalink to it that gives you that error mentioned above?

Reply:
> Hmm.. Okay. Can you do a table() of those two variables? Perhaps there's not enough samples to use both variables, in which case, that's fine, we can just explain why we are choosing one over the other.

Would you like me to do this a commit the changes or would you prefer that I post the results of `table()` under this comment for discussion purposes @cansavvy?

> I am a bit confused by this comment. Can you give me more context and/or a link?:

This was discussed through direct messaging on Slack, here's a snippet of the convo:
<img width="595" alt="Screen Shot 2020-08-11 at 2 24 00 PM" src="https://user-images.githubusercontent.com/43576623/89934423-88a5a080-dbde-11ea-874e-00466bfd48a5.png">

Reply:
> Also can you paste here the code or permalink to it that gives you that error mentioned above?

Adapted from lines 212-217:

```
# Create a `DESeqDataSet` object
dds <- DESeqDataSetFromMatrix(
  countData = df, # This is the data.frame with the counts values for all replicates in our dataset
  colData = annotation_df, # This is the data.frame with the annotation data for the replicates in the counts data.frame
  design = ~refinebio_treatment + mutation # Replace with an appropriate design variable for your analysis
)
```

I did not commit this change because of the error.

Reply:
> Hmm.. Okay. Can you do a table() of those two variables? Perhaps there's not enough samples to use both variables, in which case, that's fine, we can just explain why we are choosing one over the other.

Here you go!
`table(annotation_df$refinebio_treatment, annotation_df$mutation)`
<img width="187" alt="Screen Shot 2020-08-11 at 2 33 22 PM" src="https://user-images.githubusercontent.com/43576623/89935137-ad4e4800-dbdf-11ea-9de6-dbfa020be6ac.png">

Reply:
Yes, so the 0's here are why we can't use both variables. 

I'm a bit confused on their experimental model (like I said in the original comment I haven't looked into their paper or descriptions much), but let's not go to far in the weeds and just use one experimental variable or the other (probably `refinebio_treatment` since it is already ready to go) And then probably just add a sentence telling the user what the variable is and why we are using it? We aren't doing DE here so in the end it doesn't matter. (But still have both labeled in the heatmap).

Reply:
So basically return to how you had it 😄  (but I think this has still been a good idea to look into what our dataset has). :+1:

Reply:
Is there a reason we shouldn't explicitly blind to experimental design/use no design with `~ 1` here? (Mentioned: https://github.com/AlexsLemonade/refinebio-examples/issues/127#issuecomment-658926700)

_Side note:_ If we're in a spot where we need to include direct messages during review of a pull request, it suggests to me that the issue https://github.com/AlexsLemonade/refinebio-examples/issues/127#issuecomment-669347715 could have included more context for why the course of action was:

> Said draft PR will include a more simplified code for the clustering example notebook, that would start with the heatmap, then deal with the metadata and annotation bars one by one.